Git-commit: 40db5a537655aa0b17a82a4b5596f8d27400edd8
From: Joe Thornber <ejt@redhat.com>
Date: Fri, 27 Jul 2012 15:08:14 +0100
Subject: [PATCH] dm thin metadata: add dm_thin_changed_this_transaction
Patch-mainline: v3.6-rc1
Reference: FATE#313903

Introduce dm_thin_changed_this_transaction to dm-thin-metadata to publish a
useful bit of information we're already tracking.  This will help dm thin
decide when to commit.

Signed-off-by: Joe Thornber <ejt@redhat.com>
Signed-off-by: Mike Snitzer <snitzer@redhat.com>
Signed-off-by: Alasdair G Kergon <agk@redhat.com>
Acked-by: NeilBrown <nfbrown@suse.com>

---
 drivers/md/dm-thin-metadata.c |   16 +++++++++++++---
 drivers/md/dm-thin-metadata.h |    2 ++
 2 files changed, 15 insertions(+), 3 deletions(-)

--- linux-3.0-SLE11-SP3.orig/drivers/md/dm-thin-metadata.c
+++ linux-3.0-SLE11-SP3/drivers/md/dm-thin-metadata.c
@@ -1375,10 +1375,9 @@ static int __insert(struct dm_thin_devic
 	if (r)
 		return r;
 
-	if (inserted) {
+	td->changed = 1;
+	if (inserted)
 		td->mapped_blocks++;
-		td->changed = 1;
-	}
 
 	return 0;
 }
@@ -1421,6 +1420,17 @@ int dm_thin_remove_block(struct dm_thin_
 
 	return r;
 }
+
+bool dm_thin_changed_this_transaction(struct dm_thin_device *td)
+{
+	int r;
+
+	down_read(&td->pmd->root_lock);
+	r = td->changed;
+	up_read(&td->pmd->root_lock);
+
+	return r;
+}
 
 int dm_pool_alloc_data_block(struct dm_pool_metadata *pmd, dm_block_t *result)
 {
--- linux-3.0-SLE11-SP3.orig/drivers/md/dm-thin-metadata.h
+++ linux-3.0-SLE11-SP3/drivers/md/dm-thin-metadata.h
@@ -148,6 +148,8 @@ int dm_thin_remove_block(struct dm_thin_
 /*
  * Queries.
  */
+bool dm_thin_changed_this_transaction(struct dm_thin_device *td);
+
 int dm_thin_get_highest_mapped_block(struct dm_thin_device *td,
 				     dm_block_t *highest_mapped);
 
